src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,316 @@
+        unset($this->unset[$property]);
+        $this->exists[$property] = true;
+    }
+
+    /** @param array{bson: Document<TValue>, set: array<string, TValue>, unset: array<string, true>, codecLibrary: CodecLibrary} $data */
+    public function __unserialize(array $data): void
+    {
+        $this->bson = $data['bson'];
+        $this->set = $data['set'];
+        $this->unset = $data['unset'];
+        $this->codecLibrary = $data['codecLibrary'];
+
+        $this->exists = array_map(
+        /** @param TValue $value */
+            fn ($value): bool => true,
+            $this->set,
+        );
+
+        foreach ($this->unset as $name => $unused) {
+            $this->exists[$name] = false;
+        }
+    }
+
+    public function __unset(string $name): void
+    {
+        $this->unset[$name] = true;
+        $this->exists[$name] = false;
+        unset($this->set[$name]);
+    }
+
+    public function count(): int
+    {
+        $this->readEntireDocument();
+
+        return count(array_filter($this->exists));
+    }
+
+    /** @return Iterator<string, TValue> */
+    public function getIterator(): CallbackIterator
+    {
+        $itemIterator = new AppendIterator();
+        // Iterate through all fields in the BSON document
+        $itemIterator->append($this->bson->getIterator());
+        // Then iterate over all fields that were set
+        $itemIterator->append(new ArrayIterator($this->set));
+
+        /** @var array<string, bool> $seen */
+        $seen = [];
+
+        return new CallbackIterator(
+            // Skip keys that were unset or handled in a previous iterator
+            new CallbackFilterIterator(
+                $itemIterator,
+                /** @param TValue $current */
+                function ($current, string $key) use (&$seen): bool {
+                    return ! isset($this->unset[$key]) && ! isset($seen[$key]);
+                },
+            ),
+            /**
+             * @param TValue $value
+             * @return TValue
+             */
+            function ($value, string $key) use (&$seen) {
+                // Mark key as seen, skipping any future occurrences
+                $seen[$key] = true;
+
+                // Return actual value (potentially overridden by __set)
+                return $this->__get($key);
+            },
+        );
+    }
+
+    public function jsonSerialize(): array
+    {
+        return iterator_to_array($this->getIterator());
+    }
+
+    /** @param mixed $offset */
+    public function offsetExists($offset): bool
+    {
+        if (! is_string($offset)) {
+            return false;
+        }
+
+        return $this->__isset($offset);
+    }
+
+    /**
+     * @param mixed $offset
+     * @return TValue
+     */
+    #[ReturnTypeWillChange]
+    public function offsetGet($offset)
+    {
+        if (! is_string($offset)) {
+            trigger_error(sprintf('Undefined offset: %s', (string) $offset), E_USER_WARNING);
```

Comment:
The type check stems from a suggestion by @GromNaN who suggested that silently casting integer offsets to strings could lead to undesired behaviour. On the other hand, since I convert the offset to a string anyways this could again be confusing. I'll revisit this once more.

As for the wording, I wanted to intentionally use "offset" here as during `offsetGet` users are interacting with the document as if it was an array, so referring to an offset instead of a "property" made sense.

Reply:
> On the other hand, since I convert the offset to a string anyways this could again be confusing. I'll revisit this once more.

Is "convert the offset to a string" referring to doing so for the exception message? I assume you have to use `trigger_error()` and `E_USER_WARNING` here, but perhaps structuring the message like we do for `InvalidArgumentException::invalidType()` would make more sense?